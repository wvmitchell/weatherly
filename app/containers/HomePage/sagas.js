import { takeEvery, call, put, select } from 'redux-saga/effects';
import API from 'utils/api';
import {
  SUBMIT_LOCATION,
  ADD_CONDITIONS,
  ADD_FORECAST,
  ADD_QUERY_RESULTS,
  API_FETCH_ERROR,
} from './constants';
import {
  makeSelectLocation,
} from './selectors';

// Individual exports for testing
export function* watchSubmitLocation() {
  yield takeEvery(SUBMIT_LOCATION, submitLocation);
}

export function* submitLocation() {
  try {
    const location = yield select(makeSelectLocation());
    const forecast = yield call(API.getForecast, location);
    const conditions = yield call(API.getConditions, location);
    if (conditions.current_observation) {
      yield put({ type: ADD_CONDITIONS, conditions: conditions.current_observation });
      yield put({
        type: ADD_FORECAST,
        forecast: { simple: forecast.forecast.simpleforecast, txt: forecast.forecast.txt_forecast },
      });
    } else {
      yield put({ type: ADD_QUERY_RESULTS, results: conditions.response.results });
    }
  } catch (err) {
    yield put({ type: API_FETCH_ERROR, error: err.message });
  }
}

// All sagas to be loaded
export default [
  watchSubmitLocation,
];
