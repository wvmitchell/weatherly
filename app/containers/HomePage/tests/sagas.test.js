/**
 * Test  sagas
 */

/* eslint-disable redux-saga/yield-effects */
import { takeEvery, call, put } from 'redux-saga/effects';
import API from 'utils/api';
import { watchSubmitLocation, submitLocation } from '../sagas';
import {
  SUBMIT_LOCATION,
  ADD_CONDITIONS,
} from '../constants';

describe('HomePage Sagas', () => {
  describe('watchSubmitLocation Saga', () => {
    const generator = watchSubmitLocation();

    it('calls submitLocation', () => {
      expect(generator.next().value).toEqual(takeEvery(SUBMIT_LOCATION, submitLocation));
    });

    it('is done', () => {
      expect(generator.next()).toEqual({ done: true, value: undefined });
    });
  });

  describe('submitLocation Saga', () => {
    const generator = submitLocation();
    const mockResponse = { current_observation: { temp: 'some-temp' } };

    it('selects the location', () => {
      expect(generator.next().value).toMatchSnapshot();
    });

    it('calls the API for forecast', () => {
      expect(generator.next('80205').value).toEqual(call(API.getForecast, '80205'));
    });

    it('calls the API for conditions', () => {
      expect(generator.next('80205').value).toEqual(call(API.getConditions, '80205'));
    });

    it('updates the state with the conditions', () => {
      expect(generator.next(mockResponse).value).toEqual(put({ type: ADD_CONDITIONS, conditions: mockResponse.current_observation }));
    });

    it('is done', () => {
      expect(generator.next()).toEqual({ done: true, value: undefined });
    });
  });
});
