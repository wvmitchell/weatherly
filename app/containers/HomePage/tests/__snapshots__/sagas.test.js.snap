exports[`HomePage Sagas submitLocation Saga selects the location 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;
